src/cmd/flux-mini.py

```diff
@@ -23,6 +27,116 @@
 from flux import debugged
 
 
+def filter_dict(env, pattern, reverseMatch=True):
+    """
+    Filter out all keys that match "pattern" from dict 'env'
+
+    Pattern is assumed to be a shell glob(7) pattern, unless it begins
+    with '/', in which case the pattern is a regex.
+    """
+    if pattern.startswith("/"):
+        pattern = pattern[1::].rstrip("/")
+    else:
+        pattern = fnmatch.translate(pattern)
+    regex = re.compile(pattern)
+    if reverseMatch:
+        return dict(filter(lambda x: not regex.match(x[0]), env.items()))
+    return dict(filter(lambda x: regex.match(x[0]), env.items()))
+
+
+def get_filtered_environment(rules, environ=None):
+    """
+    Filter environment dictionary 'environ' given a list of rules.
+    Each rule can filter, set, or modify the existing environment.
+    """
+    if environ is None:
+        environ = dict(os.environ)
+    if rules is None:
+        return environ
+    for rule in rules:
+        #
+        #  If rule starts with '-' then the rest of the rule is a pattern
+        #   which filters matching environment variables from the
+        #   generated environment.
+        #
+        if rule.startswith("-"):
+            environ = filter_dict(environ, rule[1::])
+        #
+        #  If rule starts with '^', then the result of the rule is a filename
+        #   from which to read more rules.
+        #
+        elif rule.startswith("^"):
+            filename = os.path.expanduser(rule[1::])
+            with open(filename) as envfile:
+                lines = [line.strip() for line in envfile]
+                environ = get_filtered_environment(lines, environ=environ)
+        #
+        #  Otherwise, the rule is an explicit variable assignment
+        #   VAR=VAL. If =VAL is not provided then VAL refers to the
+        #   value for VAR in the current environment of this process.
+        #
+        #  Quoted shell variables are expanded using values from the
+        #   built environment, not the process environment. So
+        #   --env=PATH=/bin --env=PATH='$PATH:/foo' results in
+        #   PATH=/bin:/foo.
+        #
+        else:
+            var, *rest = rule.split("=", 1)
+            if not rest:
+                #
+                #  VAR alone with no set value pulls in all matching
+                #   variables from current environment that are not already
+                #   in the generated environment.
+                env = filter_dict(os.environ, var, reverseMatch=False)
+                for key, value in env.items():
+                    if key not in environ:
+                        environ[key] = value
+            else:
+                #
+                #  Template lookup: use jobspec environment first, fallback
+                #   to current process environment using ChainMap:
+                lookup = ChainMap(environ, os.environ)
+                environ[var] = Template(rest[0]).substitute(lookup)
```

Comment:
😍 🤤 🥳 💯 🎉 this is soooo clean! I love it!

Reply:
One minor thing (sorry for not catching this earlier), can we wrap this in a try-catch to produce a slightly nicer error message?

```
❯ flux mini run --dry-run --env-remove="*" --env='FOO=foo;${BAZ:-baz}' hostname | jq .attributes                                                                                                                               14:30:36 ()
flux-mini: ERROR: Invalid placeholder in string: line 1, col 5

herbein1@quartz2498 ~/Repositories/flux-framework/flux-core (issue#3141 ?S)
❯ FLUX_PYCLI_LOGLEVEL=0 flux mini run --dry-run --env-remove="*" --env='FOO=foo;${BAZ:-baz}' hostname | jq .attributes                                                                                                         14:30:41 ()
flux-mini: ERROR: Invalid placeholder in string: line 1, col 5
flux-mini: DEBUG: Traceback (most recent call last):
  File "/g/g0/herbein1/Repositories/flux-framework/flux-core/src/bindings/python/flux/util.py", line 136, in __call__
    main_func()
  File "/g/g0/herbein1/Repositories/flux-framework/flux-core/src/cmd/flux-mini.py", line 696, in main
    args.func(args)
  File "/g/g0/herbein1/Repositories/flux-framework/flux-core/src/cmd/flux-mini.py", line 434, in main
    jobid = self.submit(args)
  File "/g/g0/herbein1/Repositories/flux-framework/flux-core/src/cmd/flux-mini.py", line 279, in submit
    jobspec.environment = get_filtered_environment(args.env)
  File "/g/g0/herbein1/Repositories/flux-framework/flux-core/src/cmd/flux-mini.py", line 99, in get_filtered_environment
    environ[var] = Template(rest[0]).substitute(lookup)
  File "/usr/lib64/python3.6/string.py", line 130, in substitute
    return self.pattern.sub(convert, self.template)
  File "/usr/lib64/python3.6/string.py", line 127, in convert
    self._invalid(mo)
  File "/usr/lib64/python3.6/string.py", line 103, in _invalid
    (lineno, colno))
ValueError: Invalid placeholder in string: line 1, col 5
```

Reply:
Maybe something like this?

```diff
diff --git a/src/cmd/flux-mini.py b/src/cmd/flux-mini.py
index dc0d848dd..9e9e296d5 100755
--- a/src/cmd/flux-mini.py
+++ b/src/cmd/flux-mini.py
@@ -96,7 +96,11 @@ def get_filtered_environment(rules, environ=None):
                 #  Template lookup: use jobspec environment first, fallback
                 #   to current process environment using ChainMap:
                 lookup = ChainMap(environ, os.environ)
-                environ[var] = Template(rest[0]).substitute(lookup)
+                try:
+                    environ[var] = Template(rest[0]).substitute(lookup)
+                except ValueError as e:
+                    LOGGER.error("Failure to substitute env var")
+                    raise
     return environ
```

```
❯ flux mini run --dry-run --env-remove="*" --env='FOO=foo;${BAZ:-baz}' hostname | jq .attributes                                                                                                                               14:56:42 ()
flux-mini: ERROR: Failure to substitute env var
flux-mini: ERROR: Invalid placeholder in string: line 1, col 5
```

Reply:
Great idea!

Reply:
I think there is also an option to Template.substitute() to either substitute an empty string or leave the unexpanded text alone. I think I prefer generating an error as you have above.

Reply:
`Template().substitute()` also raises `KeyError` if `$VAR` is not found. I think this should still be an error, so how about:
```diff
diff --git a/src/cmd/flux-mini.py b/src/cmd/flux-mini.py
index e3a3682e6..c84928933 100755
--- a/src/cmd/flux-mini.py
+++ b/src/cmd/flux-mini.py
@@ -96,7 +96,13 @@ def get_filtered_environment(rules, environ=None):
                 #  Template lookup: use jobspec environment first, fallback
                 #   to current process environment using ChainMap:
                 lookup = ChainMap(environ, os.environ)
-                environ[var] = Template(rest[0]).substitute(lookup)
+                try:
+                    environ[var] = Template(rest[0]).substitute(lookup)
+                except ValueError as e:
+                    LOGGER.error(f"--env: Unable to substitute {rule}")
+                    raise
+                except KeyError as e:
+                    raise Exception(f"--env: Variable {e} not found in {rule}")
     return environ
```

```
ƒ(s=4,d=0,builddir) grondo@asp:~$ flux mini run --dry-run --env-remove="*" --env='FOO=foo;${BAR}' hostname
flux-mini: ERROR: --env: Variable 'BAR' not found in FOO=foo;${BAR}
ƒ(s=4,d=0,builddir) grondo@asp:~$ flux mini run --dry-run --env-remove="*" --env='FOO=foo;${BAZ:-baz}' hostname
flux-mini: ERROR: --env: Unable to substitute FOO=foo;${BAZ:-baz}
flux-mini: ERROR: Invalid placeholder in string: line 1, col 5
```

Reply:
Love it.  Also, I totally forgot that we can use f-strings now that we are 3.6+.  Thanks for that reminder and subsequent burst of joy and happiness @grondo :)

Reply:
Sadly, the version of pylint in travis doesn't like f-string:
```
W1202: Use % formatting in logging functions and pass the % parameters as arguments (logging-format-interpolation)
```